### ISSUE TITLE
nwjs 0.16 - MacOS 10.10 - nw.global.__nw_windows not correct

### ISSUE THREAD

**Issue by AndryBray** [nwjs/nw.js#5181]
My app handles the running instances of every window and it was working good with 0.14 & 0.15.
On v 0.16 I see there is some problem with `nw.global.__nw_windows`
It doesn't store always the right number of opened windows.
E.g. if I open 5 windows I see just 2 NWWindow object inside `nw.global.__nw_windows`

**Comment by AndryBray** [comment]
@rogerwang is there any other way to get all running windows (nw instances). I'm trying chrome api but it doesn't allow to access the window object (e.g. window.document)

**Comment by AndryBray** [comment]
@rogerwang just tested on Windows and it's the same... very often some window instance will not be tracked to the __nw_windows store var.

**Comment by rogerwang** [comment]
We'll provide nw.Window.getAll(). Before that you can use https://developer.chrome.com/apps/app_window#method-getAll

We'll see the issue about __nw_windows.

**Comment by AndryBray** [comment]
Yes I saw the chrome methods, but how can I access the window document object?
I see no way to interact with window scope / DOM.
E.g. I'd like to get a global var inside every window to recognize the kind of window

**Comment by AndryBray** [comment]
@rogerwang tested event with `chrome.app.window.getAll()` and it's the same... worst.. it shows only 1 AppWindow running (when actually I have 5 windows)

**Comment by AndryBray** [comment]
@rogerwang do you have any planned date for this issue solving? It could be blocking for who need to handle number of window instances. Thank you

**Comment by AndryBray** [comment]
up

**Comment by AndryBray** [comment]
@rogerwang tested the new version 0.17 (latest):
on MacOS version `nw.global.__nw_windows` now works well
on Windows version `nw.global.__nw_windows` does not hold the correct number of windows (tested on Windows 7)

**Comment by AndryBray** [comment]
Hello @rogerwang 
is this issue a 0.18 version priority? 
Sorry if I'm repetitive but I just need to know if I have to develop my own window manager.
Thank you

**Comment by rogerwang** [comment]
@AndryBray I'll try to fix it soon in 0.17. Thanks.

**Comment by rogerwang** [comment]
does it happen only when you open windows with remote URLs?

**Comment by AndryBray** [comment]
I only use remote urls. I'll try with local as well

**Comment by AndryBray** [comment]
@rogerwang I've just tested the live build version https://dl.nwjs.io/live-build/09-10-2016/nw17-4c8d707-20839a2-dacfec0-65b948e/v0.17.2/

I still have the same problem (tested on Windows 7)

**Comment by rogerwang** [comment]
Did you test with chrome.app.window.getAll()? That's what I was fixing.

**Comment by AndryBray** [comment]
@rogerwang OK. I was still using the nw.global.__windows object instead of chrome API.

Using the chrome api it seems to work. 

Just for info:
using chrome.app.window.getAll() with the sdk build, if I have dev console opened I get an array with NULL values like: [AppWindow, null, AppWindow]
Maybe the null value could be skipped automatically.

It's a pity that we still have problems with nw.window.open callback with undefined wind value.
I think https://github.com/nwjs/nw.js/issues/4188 is still there.

**Comment by AndryBray** [comment]
@rogerwang is this bugfix applied on 0.14.x version as well?

**Comment by AndryBray** [comment]
@rogerwang about the NULL value I found it's not related to dev console but chrome.app.window.getAll() will return NULL values if you try to reload the content (CMD+R) of window.
I think this shouldn't be the normal behaviour.

Tested on 0.18 beta-1 as well.

**Comment by rogerwang** [comment]
Yes, it will be applied to LTS branch too. Reopen to track this.

**Comment by AndryBray** [comment]
@rogerwang it would be nice to have a kind of roadmap for issues just to know your estimation about the problems solving or just to know if you are working on. 
Thank you, I appreciate your effort in every release you publish.

Any idea about this one especially?

**Comment by rogerwang** [comment]
Will see this before 0.17.3. I was about to fix it in 0.17.2 but was
occupied by upstream release with security updates.

**Comment by rogerwang** [comment]
Google just released another Chromium stable update in the next day after the previous one. So I'm going to release 0.17.3 soon and this issue will be likely slip into 0.17.4.

**Comment by rogerwang** [comment]
returning null is fixed in #5294

**Comment by mariomui-viscira** [comment]
> @rogerwang tested even with `chrome.app.window.getAll()` and it's the same... worst.. it shows only 1 AppWindow running (when actually I have 5 windows).
> 
> Instead if I use `chrome.windows.getAll(function(wins) { console.log(wins) })` it works but this method will not give us access to the contentWindow node (to access the document)

it does work you just have to remember that the window opens in the context of the newly focused window. if you open chrome debugger of the latest window you should see that all your windows are all open.
place this inside the callback of your xx.open(..., ( childwin) => .... 
   const oWindows = childwin.window.chrome.app.window.getAll();
          oWindows.forEach((oWindow) => {
            if (oWindow.id !== '' && oWindow.id !== modalName) {
              oWindow.close(true);
            }
          });
